[AOTI Minifier] Save EP instead of graphs

yushangdi opened the pull request:
Summary:
`repro.py` can have nested graph modules, e.g.

```
class Repro(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.true_graph_0 = GraphModule()



    def forward(self):
        true_graph_0 = self.true_graph_0
        return (true_graph_0,)
```

So dumping the string doesn’t always work.

So,
1) we use exported program in repro.py instead
2) we still dump the graph module string, but only put it in comments


We also added two flags to `minifier_launcher.py`
- `minifier-export-mode`: whether strict or non-strict export is used in the minifier
- `skip-export-error`: intermediate graphs that cannot be exported will be skipped.

Test Plan:
```
buck2 run  fbcode//caffe2/test/inductor:minifier_utils_cpu  -- -r string
python test/inductor/test_minifier.py
```

Differential Revision: D66175257




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

henrylhtsang commented on torch/_dynamo/repro/aoti.py:
```diff
@@ -53,32 +53,78 @@ def dump_to_minify(
+    elif gm is None:
+        gm = exported_program.module()
+
+    # save a graph preview using gm
```
so just to be clear, gm is just for string. exported_program is the thing that actually matters?

henrylhtsang commented on torch/_dynamo/repro/aoti.py:
```diff
@@ -53,32 +53,78 @@ def dump_to_minify(
         save_dir=subdir,
         command="minify",
-        options=options,
+        config_patches=options,
```
so it will just use strict = False?

henrylhtsang commented on torch/_dynamo/repro/aoti.py:
```diff
@@ -53,32 +53,78 @@ def dump_to_minify(
     command="run",
     accuracy=None,
     check_str=None,
+    module_in_comment=False,
```
I guess you will use it in the future?

yushangdi replied:
> I guess you will use it in the future?

yes! I think this PR is getting a little long already. 

Basically, it's for not having the module in comment in `minifier_launcher.py`, since the module can be quite large.

henrylhtsang commented on torch/_dynamo/repro/aoti.py:
```diff
@@ -290,28 +292,38 @@ def export_for_aoti_minifier(gm, tuple_inputs) -> Optional[torch.nn.Module]:
+                return None
         raise AOTIMinifierError(e) from e
-
+    return None
```
not sure what this does but okay

yushangdi replied:
> not sure what this does but okay
yeah we should never reach here.

yushangdi committed:
[AOTI Minifier] Save EP instead of graphs (#141159)

Summary:

`repro.py` can have nested graph modules, e.g.

```
class Repro(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.true_graph_0 = GraphModule()



    def forward(self):
        true_graph_0 = self.true_graph_0
        return (true_graph_0,)
```

So dumping the string doesn’t always work.

So,
1) we use exported program in repro.py instead
2) we still dump the graph module string, but only put it in comments


We also added two flags to `minifier_launcher.py`
- `minifier-export-mode`: whether strict or non-strict export is used in the minifier
- `skip-export-error`: intermediate graphs that cannot be exported will be skipped.

Test Plan:
```
buck2 run  fbcode//caffe2/test/inductor:minifier_utils_cpu  -- -r string
python test/inductor/test_minifier.py
```

Reviewed By: henrylhtsang

Differential Revision: D66175257
